File: src/InvertedIndex.ts
type char = string;

export class InvertedIndex<T> {
  private map: Map<char, number[]>;
  private items: T[];

  constructor() {
    this.map = new Map();
    this.items = [];
  }

  add(item: T, text: string) {
    const index = this.items.length;
    this.items.push(item);

    for (const ch of text) {
      let indexes = this.map.get(ch);
      if (indexes === undefined) {
        indexes = [];
        this.map.set(ch, indexes);
      }
      indexes.push(index);
    }
  }

  get(text: string) {
    let result = this.map.get(text[0]);
    if (result === undefined || result.length === 0) {
      return [];
    }
    for (let i = 1; i < text.length; i++) {
      const indexes = this.map.get(text[i]);
      if (indexes === undefined) {
        return [];
      }
      result = intersect(result, indexes);
      if (result.length === 0) {
        return [];
      }
    }
    return result.map((index) => this.items[index]);
  }

  getItems() {
    return this.items;
  }
}

function intersect(a: number[], b: number[]): number[] {
  if (a.length === 0 || b.length === 0) {
    return [];
  }
  if (a[0] > b[b.length - 1]) {
    return [];
  }
  if (b[0] > a[a.length - 1]) {
    return [];
  }
  const result = [];
  for (let i = 0, j = 0; i < a.length && j < b.length; ) {
    if (a[i] === b[j]) {
      result.push(a[i]);
      i++;
      j++;
    } else if (a[i] < b[j]) {
      i++;
    } else {
      j++;
    }
  }
  return result;
}

File: internal/inverted_index.go
package util

import (
	"github.com/Lechros/hangul_regexp"
	"strconv"
	"unicode/utf8"
)

type HangulInvertedIndex[T any] struct {
	items []T
	// key: rune, value: slice of idx of data
	index map[rune][]int
}

func (ii *HangulInvertedIndex[T]) Init() {
	ii.items = make([]T, 0)
	ii.index = make(map[rune][]int)
}

func (ii *HangulInvertedIndex[T]) Add(item T, key string) {
	itemIdx := len(ii.items)
	ii.items = append(ii.items, item)
	visited := make(map[rune]struct{}, len(key))
	for _, ch := range key {
		if _, ok := visited[ch]; !ok {
			ii.index[ch] = append(ii.index[ch], itemIdx)
			visited[ch] = struct{}{}
		}
	}
}

func (ii *HangulInvertedIndex[T]) FindAll(key string) []T {
	if len(key) <= 1 {
		panic("key should have length > 1 but was " + strconv.Itoa(len(key)))
	}

	var idxList []int = nil
	for i, ch := range key {
		if i+utf8.RuneLen(ch) == len(key) {
			break
		}
		if hangul_regexp.CanBeChoseongOrJongseong(ch) {
			continue
		}
		if idxList == nil {
			idxList = make([]int, len(ii.index[ch]))
			copy(idxList, ii.index[ch])
		} else {
			idxList = intersect(idxList, ii.index[ch])
		}
	}
	if idxList == nil {
		return nil
	}
	result := make([]T, len(idxList))
	for i, idx := range idxList {
		result[i] = ii.items[idx]
	}
	return result
}

func intersect(dst, other []int) []int {
	count := 0
	i, j := 0, 0
	for i < len(dst) && j < len(other) {
		if dst[i] < other[j] {
			i++
		} else if dst[i] > other[j] {
			j++
		} else {
			dst[count] = dst[i]
			count++
			i++
			j++
		}
	}
	return dst[:count]
}
